### ISSUE TITLE
[Test] Client reboots between requests

### ISSUE THREAD

**Issue by sylfabre** [api-platform/api-platform#1668]
**Description**  
I'm using the AliceBundle for my fixtures, and this bundle relies on MySQL transactions to get a clean set of data between tests:
* When the test kernel is booted, a MySQL transaction starts
* When the test kernel shutdowns, the current MySQL transaction is rolled back

Thanks to the ApiTestCase, it is very easy to create a client to run API Platform requests in the context of functional tests.
During its creation, a Symfony kernel is booted and is "injected" in the test client.
=> It means that the test and "server" answering the requests share the same kernel so share the same container so share the same Doctrine connections.

But by default, once the Symfony Client has served its a request, it will reboot before serving a next one.

In my case, a second request will use another container so another Doctrine connection which is unaware of the current MySQL transaction (as it is isolated by-design) of the test scope.

This behavior of Symfony is not documented and unexpected which makes poor developper experience when using the `ApiTestCase`.

An easy fix is to call the `\Symfony\Bundle\FrameworkBundle\Client::disableReboot()` method.

I may also be missing a point in the current implementation.

Would you accept a PR to call `disableReboot()` in the `createClient()` method?

**Example**

<?php

// Method part of a test suite using the AliceBundle
function test1()
{
    $client = self::createClient; // This begins a MySQL transaction

    $user = new User();
    $em = self::$container->get(EntityManagerInterface::class);
    $em->persist($user);
    $em->flush();

    $client->request('GET', '/user/' . $user->getId()); // HTTP 200 => OK
    $client->request('GET', '/user/' . $user->getId()); // HTTP 404 => Unexpected
}

// Method part of a test suite using the AliceBundle
function test2()
{
    $client = self::createClient; // This begins a MySQL transaction
    $client->disableReboot();

    $user = new User();
    $em = self::$container->get(EntityManagerInterface::class);
    $em->persist($user);
    $em->flush();

    $client->request('GET', '/user/' . $user->getId()); // HTTP 200 => OK
    $client->request('GET', '/user/' . $user->getId()); // HTTP 200 => OK
}

**Comment by soyuka** [comment]
We should at least document that solution. When you disable the reboot, the isolation by-test is still working?

**Comment by sylfabre** [comment]
> When you disable the reboot, the isolation by-test is still working?

Yes it is because the Alice bundle rolls back the MySQL transaction when a call to `ensureKernelShutdown()` is made.

And this method is called during the `tearDown()` call.

**Comment by soyuka** [comment]
WDYT about fixing this directly in symfony's client though?

**Comment by sylfabre** [comment]
@soyuka Because I am not aware of all usages of the Symfony's KernelBrowser: it may have use cases where rebooting after each request is legitimate. Changing it would break these use cases.

I'll still open an issue at Symfony so experts can discuss it.

**Comment by soyuka** [comment]
I've given some more thoughts about this, I don't think that `disableReboot` should be the default but it should defiitely be documented.

**Comment by carlobeltrame** [comment]
I can confirm this issue still exists in Symfony 5. Since it looks like this is a complicated topic and there is no consensus yet where the fix should be implemented between API platform, Symfony and the alice bundle, I have created a [documentation PR](https://github.com/api-platform/docs/pull/1363) to at least document the current situation until this issue is properly resolved.
